### PR TITLE
Refactor mounting code on all platforms and make normal mount failure…

### PR DESCRIPTION
**Pull request by taruti**
…s non-fatal

Also fix race conditions in the original code and create a new abstraction in libfs.

### CODE REVIEW

File: libfuse/start.go [hunk 1]
```diff
@@ -42,52 +46,54 @@ func Start(mounter Mounter, options StartOptions, kbCtx libkbfs.Context) *libfs.
...
+				mi.Done()
+				return libfs.MountError(err.Error())
+			}
 		}
 	}
+	mi.Wait()
+	return nil
+}
+
+func startMounting(config libkbfs.Config, options StartOptions,
```
Comment (songgao) [inline libfuse/start.go]: This is only called by `func Start`; would you mind moving this above it?

File: libfuse/start.go [hunk 2]
```diff
@@ -42,52 +46,54 @@ func Start(mounter Mounter, options StartOptions, kbCtx libkbfs.Context) *libfs.
...
+	log logger.Logger, mi *libfs.MountInterrupter) error {
+	log.Debug("Mounting: %q", options.MountPoint)
+	c, err := fuseMount(options)
+	if err != nil {
+		return err
+	}
+	mi.SetOnceFun(func() { fuseUnmount(options) })
```
Comment (songgao) [inline libfuse/start.go]: There could be a race condition here where `mi.Done` was called after the `fuseMount` above, but before the `SetOnceFun` here manages to grab to lock, and the mount point will never be unmounted.

Perhaps we can have a channel `doneCalled` that `mi.Done` close in a separate`Once` wrapper, before the `mi.once.Done` is called, and `SetOnceFun` would call the Fun (through `mi.once.Done`) if the channel is already closed?

File: libfs/mount_interrupter.go [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+// license that can be found in the LICENSE file.
+
+package libfs
+
+import (
+	"sync"
+)
+
+// MountInterrupter is for managing mounts with cancelation.
+type MountInterrupter struct {
```
Comment (songgao) [inline libfs/mount_interrupter.go]: This type isn't really mounter specific, so I think a more generic name like `LazyOnceWaitable` would work as well. But I'm OK with the `MountInterrupter` too; just thought I'd point it out.

**Comment by taruti** [comment]
I ended up rewriting some parts to avoid the "interrupt just after mount" and "interrupt before mount" issues.

**Pull request merged**